A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+be recorded at the beginning of the call, while a queue-size may be reported at any time during the call life-cycle.
+The data is dumped and translated to `OrcaLoadReport` at the end of each RPC by the interceptor, 
+and sent out as the trailing metadata.
+
+Note that reporting utilization metrics per-query is only useful in high QPS systems. 
+In such systems, the application that measures utilization data should consider sampling the data 
+to avoid performance degradation due to high rates of syscalls.
+
+
+### Out of Band Metrics Reporting
+For the out-of-band reporting, it requires a separate load reporting server running along with the 
+backend server for as long as the server is alive.
+
+At the client side, typically an LB policy, opens a stream on each
+established connection to the backend server to request the load report, and then periodically receiving
+metrics data from the streaming server for each backend.
+The report request should specify metrics reporting interval, the lower bound of which should be capped 
+to a server configurable value, and there is no upper bound unless it is needed in the future.
+
+#### The Server API 
+Provide builtin implementation of `OpenRCAService` streaming service defined by ORCA in each supported language.
+Server applications are free to add such a load reporting server
+in a normal way that they register a service.
+
+The service holds the utilization metrics data in key/value pairs using a hashmap. 
+Expose set-style APIs for user applications to update the metrics data, pseudocode below:
+
+```java
+class OrcaServiceImpl implements OpenRcaService {
+
+// Allow configuring minimum report interval. If not configured or badly configured non-positive, 
+// the default is 30s.
+// In the future, it might be useful to let users specify minimum report interval and default 
+// report interval separately. If the default report interval is not configured or badly configured 
+// non-positive, the default is 1 min. 
+// public OpenRcaServiceImpl(long minInterval, TimeUnit timeUnit, long 
+// defaultInterval, TimeUnit)
+public OrcaServiceImpl(long minReportInterval, TimeUnit unit);
+
+// Update the metrics value corresponding to the specified key.
+void setUtilizationMetric(String key, double value);
+
+// Replace the whole metrics data using the specified map.
+void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+// Remove the metrics data entry corresponding to the specified key.
+void deleteUtilizationMetric(String key);
+
+// Update the CPU utilization metrics data.
+void setCPUUtilizationMetric(double value);
+
+// Clear the CPU utilization metrics data.
+void deleteCPUUtilizationMetric();
+
+// Update the memory utilization metrics data.
+void setMemoryUtilizationMetric(double value);
+
+// Clear the memory utilization metrics data.
+void deleteMemoryUtilizationMetric();
+}
+```
+
+All the methods above should be thread safe. 
```

Comment:
As above, thread-safety may be something to mention for Java, but it's over-specifying the language-agnostic design.